README.md
# TJ-like Agenda

<!-- TOC -->
* [Overview](#overview)
* [Installation](#installation)
* [Usage](#usage)
* [TODOs](#todos)
* [Contribution](#contribution)
* [Licence](#license)
<!-- TOC -->
## Overview

Pull Telegram (more platforms to be added) publications and pick up the most trending ones.

This application allows you to collect publications from pre-configured telegram (ATM) channels with no authentication required, neither a user account nor a service key. Scraping is currently done through the publicly available UI and is mainly intended to pull posts' IDs that can be used in a widget rather than downloading all the contents of the post, though it could be reworked in the future. 

Scope of application might primarily be social news aggregators such as TJournal (good night, sweet prince!) and the like which could make use of this software as a self-hosted microservice.

## Installation
- [Download](https://github.com/alexeyvy/tjlike-agenda/releases/latest) the latest archive with the binary for your OS/CPU
- Extract the archive to any folder and navigate to it
- Make sure [config.yaml](config.yaml) is placed in the working directory before running binary
- Alter `config.yaml` so that it reflects your preferred channel pool to gather publications from

## Usage
- Run the binary `./tjlike-agenda` and give it few moments to scrape publications (watch the output to track its progress)
- Upon at least 1 traversal completes, you can find the collected reposts JSON-serialized in the DB which by default is simply a file in the working directory that is called `tjlike_agenda_db.txt` and spawned/appended automatically
- However, you don't want dealing with the raw DB file which may further be replaced with another storage implementation, instead use the REST API endpoint described in [api.yml](api.yml), as follows:
```
curl localhost:35971/reposts
```
- Note that this endpoint in the current implementation isn't idempotent meaning all returned reposts evaporate from the DB as soon as the endpoint is called 
## TODOs
- Dockerize
- Better strategy on cross-channel selection
- Channel priorities
- Allow to opt for not purging read reposts
- Introduce webhooks or a queue transport to deliver reposts so periodical pulling is eliminated (?)

## Contribution
Current implementation only takes into account publication's view counter compared to previous publications' view counters. The more the counter deviates from preceding ones, the more trending it's recognized as trending within the channel which is pretty straightforward.
Whatever insights you have as to how to improve the algorithm, feel free to contribute or discuss.

Contribution to other parts are also welcome.


## License
[MIT](LICENSE)

CHANGELOG.md
# Change Log

## [0.1.0] - 2022-09-04

Initial release


infra/repost/store_test.go
package repost

import (
	"github.com/alexeyvy/tjlike-agenda/domain"
	"testing"
	"time"
)

func TestInMemoryInsertAndRead(t *testing.T) {
	s := NewInMemoryStore()
	s.insert(&entry{
		R: domain.NewRepost(domain.NewPublication("platform/id", 10300, time.Now()), time.Now(), domain.SuggestionRate(5)),
	})
	s.insert(&entry{
		R: domain.NewRepost(domain.NewPublication("platform/id2", 10300, time.Now()), time.Now(), domain.SuggestionRate(5)),
	})

	entries := make([]entry, 0)
	s.walkAll(func(e *entry) {
		entries = append(entries, *e)
	})
	if len(entries) != 2 {
		t.Errorf(
			"Expected 2 entries in the store, got %d", len(entries),
		)
	}
}

func TestInMemoryIdSequence(t *testing.T) {
	s := NewInMemoryStore()
	e1 := entry{
		R: domain.NewRepost(domain.NewPublication("platform/id", 10300, time.Now()), time.Now(), domain.SuggestionRate(5)),
	}
	s.insert(&e1)
	e2 := entry{
		R: domain.NewRepost(domain.NewPublication("platform/id2", 10300, time.Now()), time.Now(), domain.SuggestionRate(5)),
	}
	s.insert(&e2)

	s.walkAll(func(e *entry) {
		if e.Id != 1 && e.Id != 2 {
			t.Errorf(
				"Expected sequential IDs 1 and 2, got %d", e.Id,
			)
		}
	})
}

func TestInMemoryDeletion(t *testing.T) {
	s := NewInMemoryStore()

	e1 := entry{
		R: domain.NewRepost(domain.NewPublication("platform/id", 10300, time.Now()), time.Now(), domain.SuggestionRate(5)),
	}
	s.insert(&e1)
	e2 := entry{
		R: domain.NewRepost(domain.NewPublication("platform/id2", 10300, time.Now()), time.Now(), domain.SuggestionRate(5)),
	}
	s.insert(&e2)

	s.delete(e1)

	entries := make([]entry, 0)
	s.walkAll(func(e *entry) {
		entries = append(entries, *e)
	})
	if len(entries) != 1 {
		t.Errorf(
			"Expected 1 entry left in the store after deleting 1, got %d", len(entries),
		)
	}

	s.delete(e2)

	entries = make([]entry, 0)
	s.walkAll(func(e *entry) {
		entries = append(entries, *e)
	})
	if len(entries) != 0 {
		t.Errorf(
			"Expected no entry left in the store after deleting 2, got %d", len(entries),
		)
	}
}


infra/repost/service_test.go
package repost

import (
	"github.com/alexeyvy/tjlike-agenda/domain"
	"testing"
	"time"
)

func TestSaveAndPickup(t *testing.T) {
	t.Parallel()

	s := NewService(NewInMemoryStore())

	postedAt, _ := time.Parse(time.RFC822, "02 Jan 06 15:04 MST")
	repost, err := s.Repost(domain.NewPublication("platform/id", 10300, postedAt), domain.SuggestionRate(5))
	if err != nil {
		t.Errorf(
			"Repost() threw an error: " + err.Error(),
		)
	}
	if repost == (domain.Repost{}) {
		t.Errorf(
			"Returned repost is empty",
		)
	}

	reposts := s.PickUpMostTrending(false)
	if len(reposts) != 1 {
		t.Errorf(
			"Picking up saved repost failed",
		)
	}
	reposts = s.PickUpMostTrending(true)
	if len(reposts) != 1 {
		t.Errorf(
			"Dry mode failed, cannot pick up repetatively",
		)
	}
	reposts = s.PickUpMostTrending(false)
	if len(reposts) != 0 {
		t.Errorf(
			"%d entries left off after picked up in wet mode", len(reposts),
		)
	}
}

func TestPurgeIrrelevant(t *testing.T) {
	t.Parallel()

	var currentTime string
	now = func() time.Time { a, _ := time.Parse(time.RFC822, currentTime); return a }

	s := NewService(NewInMemoryStore())

	currentTime = "02 Jan 06 15:06 MST"
	postedAt, _ := time.Parse(time.RFC822, "02 Jan 06 15:04 MST")
	_, _ = s.Repost(domain.NewPublication("platform/id", 10300, postedAt), domain.SuggestionRate(5))

	currentTime = "06 Jan 06 15:06 MST"
	err := s.PurgeIrrelevant()
	if err == nil {
		t.Errorf(
			"PurgeUnreadError expected",
		)
	}
}


infra/scraping/telegram_test.go
package scraping

import (
	"testing"
)

func TestDehumanitizeViewNumber(t *testing.T) {
	t.Parallel()
	type table struct {
		input string
		want  int
	}

	tables := []table{
		{input: "1K", want: 1000},
		{input: "1.5K", want: 1500},
		{input: "12.5K", want: 12500},
		{input: "4.91M", want: 4910000},
		{input: "0M", want: 0},
		{input: "0", want: 0},
		{input: "55", want: 55},
	}

	for _, test := range tables {
		result := dehumanizeViewNumber(test.input)

		if result != test.want {
			t.Errorf("Dehumanized incorrectly on input %s, got: %d, want: %d.", test.input, result, test.want)
		}
	}
}

func TestConvertTgPublicationToAbstractPublication(t *testing.T) {
	t.Parallel()
	tm := telegramPublication{"tg/msgid", "10.5K", "2022-08-29T11:41:26+00:00"}
	generalized := tm.generalize()

	if generalized.PostedAt.String() != "2022-08-29 11:41:26 +0000 UTC" {
		t.Errorf(
			"PostedAt generalized incorrectly, got %s, expected: %s",
			generalized.PostedAt.String(),
			"2022-08-29 11:41:26 +0000 UTC",
		)
	}
}


domain/selection.go
package domain

import (
	"errors"
	log "github.com/sirupsen/logrus"
)

type (
	SuggestionRate      float64
	simpleLocalSelector struct{}
)

func (s *simpleLocalSelector) SelectPublication(publications []Publication) (Publication, SuggestionRate) {
	var (
		maxRate                float64
		rate                   float64
		topPublication         Publication
		prevOverweight         = 1.0
		prevPrevOverweight     = 1.0
		prevPrevPrevOverweight = 1.0
	)

	for key, publication := range publications {
		if key <= 2 {
			continue
		}
		prevOverweight = float64(publication.ViewAmount) / float64((publications[key-1]).ViewAmount)
		prevPrevOverweight = float64(publication.ViewAmount) / float64((publications[key-2]).ViewAmount)
		prevPrevPrevOverweight = float64(publication.ViewAmount) / float64((publications[key-3]).ViewAmount)
		if prevOverweight <= 1 {
			continue
		}
		rate = prevOverweight*1.5 + prevPrevOverweight*1.2 + prevPrevPrevOverweight
		if rate > maxRate {
			log.WithFields(log.Fields{
				"id":                 publication.Id,
				"rate":               rate,
				"prevOverweight":     prevOverweight,
				"prevPrevOverweight": prevPrevOverweight,
			}).Debug("New leader of channel")

			maxRate = rate
			topPublication = publication
		}
	}

	if maxRate == 0 {
		return Publication{}, SuggestionRate(0)
	}
	return topPublication, SuggestionRate(maxRate)
}
func NewLocalSelector() *simpleLocalSelector {
	return &simpleLocalSelector{}
}

type LocalSelector interface {
	SelectPublication([]Publication) (Publication, SuggestionRate)
}

type SimpleGlobalSelector struct {
	LocalSelector LocalSelector
}

func NewGlobalSelector(localSelector LocalSelector) *SimpleGlobalSelector {
	return &SimpleGlobalSelector{localSelector}
}

var ErrExhausted = errors.New("all channels exhausted")

const suggestionRateThreshold = 4

func (s *SimpleGlobalSelector) SelectPublication(
	candidates map[Channel][]Publication,
	exists func(publication Publication) bool,
) (Publication, SuggestionRate, error) {
	maxSuggestionRate := 0.0
	var selectedPublication Publication

	// @TODO implement cross-channel selection rules, including chanel priorities

	for _, channelPublications := range candidates {
		msg, suggestionRate := s.LocalSelector.SelectPublication(channelPublications)
		if exists(msg) {
			continue
		}
		if suggestionRate < suggestionRateThreshold {
			continue
		}
		if float64(suggestionRate) > maxSuggestionRate {
			maxSuggestionRate = float64(suggestionRate)
			selectedPublication = msg
		}
	}
	if maxSuggestionRate == 0 {
		return Publication{}, SuggestionRate(0), ErrExhausted
	}

	return selectedPublication, SuggestionRate(maxSuggestionRate), nil
}


infra/repost/store.go
package repost

import (
	"encoding/json"
	"errors"
	"fmt"
	"github.com/alexeyvy/tjlike-agenda/domain"
	"os"
	"sync"
)

type entry struct {
	R                    domain.Repost
	RetrievedAtLeastOnce bool
	Id                   int
}

type InMemoryStore struct {
	entries map[int]*entry
	nextId  int
	sync.RWMutex
}

func NewInMemoryStore() *InMemoryStore {
	return &InMemoryStore{make(map[int]*entry, 0), 1, sync.RWMutex{}}
}

func (s *InMemoryStore) insert(e *entry) {
	e.Id = s.nextId
	s.entries[e.Id] = e
	s.nextId += 1
}
func (s *InMemoryStore) walkAll(handle func(e *entry)) {
	for k := range s.entries {
		handle(s.entries[k])
	}
}
func (s *InMemoryStore) delete(e entry) {
	delete(s.entries, e.Id)
}

type FileStore struct {
	InMemoryStore
	path string
}

func NewFileStore(path string) *FileStore {
	return &FileStore{*NewInMemoryStore(), path}
}

type jsonRepresentation struct {
	Entries map[int]*entry
	NextId  int
}

var ErrDBNotInited = errors.New("cannot initialize store as the source does not exist")

func (fs *FileStore) Pull() error {
	if _, err := os.Stat(fs.path); errors.Is(err, os.ErrNotExist) {
		return ErrDBNotInited
	}
	data, err := os.ReadFile(fs.path)
	var jr jsonRepresentation
	err = json.Unmarshal(data, &jr)
	if err != nil {
		return fmt.Errorf("cannot unmarshal store: %w", err)
	}
	fs.entries = jr.Entries
	fs.nextId = jr.NextId
	return nil
}
func (fs *FileStore) Push() error {
	jr := jsonRepresentation{fs.entries, fs.nextId}
	jsonEncoded, _ := json.Marshal(jr)
	err := os.WriteFile(fs.path, jsonEncoded, 0644)
	if err != nil {
		return fmt.Errorf("cannot write DB file: %w", err)
	}
	return nil
}


infra/repost/service.go
package repost

import (
	"github.com/alexeyvy/tjlike-agenda/domain"
	"time"
)

type Store interface {
	insert(*entry)
	walkAll(func(e *entry))
	delete(entry)
}
type LockableStore interface {
	RLock()
	RUnlock()
	Lock()
	Unlock()
}
type PersistentStore interface {
	Pull() error
	Push() error
}

type service struct {
	s Store
}

func NewService(store Store) *service {
	persistentStore, isPersistentStore := store.(PersistentStore)
	if isPersistentStore {
		if err := persistentStore.Pull(); err != nil {
			if err != ErrDBNotInited {
				panic("cannot create repost service due to failed persistentStore.Pull(): " + err.Error())
			}
		}
	}

	return &service{store}
}

var now = time.Now

const hoursBeforePurge = 48

type PersistDBFailed struct {
	originalError error
}

func (e *PersistDBFailed) Error() string {
	return "cannot persist DB: " + e.originalError.Error()
}

func (r *service) Repost(publication domain.Publication, rate domain.SuggestionRate) (domain.Repost, error) {
	repost := domain.NewRepost(publication, now(), rate)
	dbEntry := &entry{repost, false, 0}

	lockableStore, isLockable := r.s.(LockableStore)
	if isLockable {
		lockableStore.Lock()
		defer lockableStore.Unlock()
	}
	r.s.insert(dbEntry)

	persistentStore, isPersistentStore := r.s.(PersistentStore)
	if isPersistentStore {
		if err := persistentStore.Push(); err != nil {
			return repost, &PersistDBFailed{err}
		}
	}

	return repost, nil
}

func (r *service) PickUpMostTrending(clearUp bool) []domain.Repost {
	reposts := make([]domain.Repost, 0)

	lockableStore, isLockable := r.s.(LockableStore)
	if isLockable {
		lockableStore.RLock()
		defer lockableStore.RUnlock()
	}

	r.s.walkAll(func(e *entry) {
		if e.RetrievedAtLeastOnce == false {
			reposts = append(reposts, e.R)
			if clearUp {
				e.RetrievedAtLeastOnce = true
			}
		}
	})

	return reposts
}

type PurgeUnreadError struct{}

func (e PurgeUnreadError) Error() string {
	return "storage can't be emptied, as there are reposts that have not yet been pulled"
}

func (r *service) PurgeIrrelevant() error {
	repostedAtThreshold := now().Add(-1 * hoursBeforePurge * time.Hour)
	entriesForDeletion := make([]entry, 0)
	var unreadFound bool

	lockableStore, isLockable := r.s.(LockableStore)
	if isLockable {
		lockableStore.Lock()
		defer lockableStore.Unlock()
	}

	r.s.walkAll(func(e *entry) {
		if e.R.RepostedAt.Before(repostedAtThreshold) {
			if e.RetrievedAtLeastOnce == false {
				unreadFound = true
			}
			entriesForDeletion = append(entriesForDeletion, *e)
		}
	})

	if unreadFound {
		return &PurgeUnreadError{}
	}

	for _, e := range entriesForDeletion {
		r.s.delete(e)
	}

	persistentStore, isPersistentStore := r.s.(PersistentStore)
	if isPersistentStore {
		if err := persistentStore.Push(); err != nil {
			return &PersistDBFailed{err}
		}
	}

	return nil
}

func (r *service) ExistsForPublication(publication domain.Publication) bool {
	var exists bool

	lockableStore, isLockable := r.s.(LockableStore)
	if isLockable {
		lockableStore.RLock()
		defer lockableStore.RUnlock()
	}

	r.s.walkAll(func(e *entry) {
		if e.R.Pub.Id == publication.Id {
			exists = true
		}
	})

	return exists
}


infra/scraping/telegram.go
package scraping

import (
	"errors"
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"github.com/alexeyvy/tjlike-agenda/domain"
	"net/http"
	"strconv"
	"time"
)

type TelegramScraper struct {
}

const maxMsgNum = 20

type telegramPublication struct {
	id         string
	viewAmount string
	postedAt   string
}

func dehumanizeViewNumber(humanized string) int {
	var viewAmount int

	switch humanized[len(humanized)-1:] {
	case "K":
		f, _ := strconv.ParseFloat(humanized[:len(humanized)-1], 64)
		viewAmount = int(f * 1000)
	case "M":
		f, _ := strconv.ParseFloat(humanized[:len(humanized)-1], 64)
		viewAmount = int(f * 1000000)
	default:
		viewAmount, _ = strconv.Atoi(humanized)
	}
	return viewAmount
}

func (tm telegramPublication) generalize() domain.Publication {
	postedAt, _ := time.Parse(time.RFC3339, tm.postedAt)
	return domain.NewPublication(tm.id, dehumanizeViewNumber(tm.viewAmount), postedAt.UTC())
}

func (s *TelegramScraper) ScrapeRecentPublications(channel domain.Channel) ([]domain.Publication, error) {
	unformattedPublications, err := s.scrapeRecentPublications(channel.Id)
	if err != nil {
		return nil, err
	}

	formattedPublications := make([]domain.Publication, 0, maxMsgNum)
	for publicationKey := range unformattedPublications {
		formattedPublications = append(formattedPublications, unformattedPublications[publicationKey].generalize())
	}
	return formattedPublications, nil
}

func (s *TelegramScraper) scrapeRecentPublications(channelId string) ([]telegramPublication, error) {
	url := fmt.Sprintf("https://t.me/s/%s", channelId)
	res, err := http.Get(url)
	if err != nil {
		return nil, fmt.Errorf("HTTP request failed: %w", err)
	}
	defer res.Body.Close()
	if res.StatusCode != 200 {
		return nil, errors.New("HTTP request failed")
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, fmt.Errorf("cannot parse HTTP response: %w", err)
	}

	var publications []telegramPublication

	container := doc.Find("div.tgme_widget_message_wrap")
	if container.Length() == 0 {
		return nil, errors.New("no publications found in div.tgme_widget_message_wrap. this may be private or non-existent channel")
	}

	var e error
	container.Each(func(i int, s *goquery.Selection) {
		idContainer := s.Find("div.tgme_widget_message")
		var id string
		if idContainer.Length() == 1 {
			var exists bool
			id, exists = idContainer.Attr("data-post")
			if !exists {
				e = fmt.Errorf("one of publications has no ID")
				return
			}
		} else {
			e = fmt.Errorf("container ID not found")
			return
		}

		viewsContainer := s.Find(".tgme_widget_message_views")
		// this is an expected issue, skip one without views
		if viewsContainer.Length() == 0 {
			return
		}
		viewAmount := viewsContainer.Text()

		postedAtContainer := s.Find(".tgme_widget_message_date time")
		var postedAt string
		if postedAtContainer.Length() == 1 {
			var exists bool
			postedAt, exists = postedAtContainer.Attr("datetime")
			if !exists {
				e = fmt.Errorf("one of publications has no POSTED AT")
				return
			}
		} else {
			e = fmt.Errorf("container POSTED AT not found")
			return
		}

		publications = append(publications, telegramPublication{id: id, viewAmount: viewAmount, postedAt: postedAt})
	})
	if e != nil {
		return publications, e
	}

	return publications, nil
}


main.go
package main

import (
	"encoding/json"
	domain "github.com/alexeyvy/tjlike-agenda/domain"
	"github.com/alexeyvy/tjlike-agenda/infra/repost"
	"github.com/alexeyvy/tjlike-agenda/infra/scraping"
	"github.com/gorilla/mux"
	log "github.com/sirupsen/logrus"
	"gopkg.in/yaml.v3"
	"io/ioutil"
	"net/http"
	"strconv"
	"sync"
	"time"
)

type Scraper interface {
	ScrapeRecentPublications(domain.Channel) ([]domain.Publication, error)
}

type scraperPoolElement struct {
	platformId string
	s          Scraper
	config     ScraperConfigEntry
}

type ScraperConfigEntry struct {
	Parallel                 bool
	Frequency                int
	PauseBetweenSyncChannels int `yaml:"pause_between_sync_channels"`
	Channels                 []string
}
type preferences struct {
	Scrapers struct {
		Telegram ScraperConfigEntry
	}
}

func initPreferences() preferences {
	log.Info("Reading YAML configuration at config.yaml")

	yamlFile, err := ioutil.ReadFile("config.yaml")
	if err != nil {
		log.Fatalf("YAML config reading error: #%v ", err)
	}
	var preferences preferences
	err = yaml.Unmarshal(yamlFile, &preferences)
	if err != nil {
		log.Fatalf("YAML config unmarshal error: %v", err)
	}
	return preferences
}

type GlobalSelector interface {
	SelectPublication(
		candidates map[domain.Channel][]domain.Publication,
		exists func(publication domain.Publication) bool,
	) (domain.Publication, domain.SuggestionRate, error)
}
type RepostWriterService interface {
	Repost(domain.Publication, domain.SuggestionRate) (domain.Repost, error)
	ExistsForPublication(domain.Publication) bool
}

func main() {
	preferences := initPreferences()

	store := repost.NewFileStore("tjlike_agenda_db.txt")
	runApi(store)

	scraperPool := []scraperPoolElement{
		{
			"telegram",
			&scraping.TelegramScraper{},
			preferences.Scrapers.Telegram,
		},
	}

	var localSelector domain.LocalSelector
	localSelector = domain.NewLocalSelector()
	var selector GlobalSelector
	selector = domain.NewGlobalSelector(localSelector)

	var repostWriter RepostWriterService
	repostWriter = repost.NewService(store)

	for _, scraperPoolEl := range scraperPool {
		scraperPoolEntry := scraperPoolEl
		go func() {
			for {
				collectedPublications := make(map[domain.Channel][]domain.Publication, 1000)
				var (
					scraperWg sync.WaitGroup
					msgMutex  sync.Mutex
				)
				log.Debugf(
					"Preparing to scrape %d channels for platform %s",
					len(scraperPoolEntry.config.Channels),
					scraperPoolEntry.platformId,
				)
				for _, src := range scraperPoolEntry.config.Channels {
					f := func(src string) {
						if scraperPoolEntry.config.Parallel {
							defer scraperWg.Done()
						}
						channel := domain.NewChannel(src)
						publications, err := scraperPoolEntry.s.ScrapeRecentPublications(channel)
						if err != nil {
							log.Errorf("scraping failed on channel %s platform %s: %s", channel.Id, scraperPoolEntry.platformId, err.Error())
							return
						}
						msgMutex.Lock()
						collectedPublications[channel] = publications
						msgMutex.Unlock()
					}
					if scraperPoolEntry.config.Parallel {
						scraperWg.Add(1)
						go f(src)
					} else {
						f(src)
						time.Sleep(time.Second * time.Duration(scraperPoolEntry.config.PauseBetweenSyncChannels))
					}
				}
				scraperWg.Wait()

				log.Debugf("All channels finished for platform %s", scraperPoolEntry.platformId)

				best, suggestionRate, err := selector.SelectPublication(collectedPublications, repostWriter.ExistsForPublication)
				if err == domain.ErrExhausted {
					log.Infof("No trending publications for platform %s so far", scraperPoolEntry.platformId)
				}
				if err == nil {
					if _, err := repostWriter.Repost(best, suggestionRate); err != nil {
						log.Errorf("Repost succeeded, however, there was an error when persisting it in the DB: %s", err)
					}

					log.Infof("Picked most trending publication %s for platform %s", best.Id, scraperPoolEntry.platformId)
				}
				log.Debugf(
					"Sleeping %d seconds before next traversal for platform %s",
					scraperPoolEntry.config.Frequency,
					scraperPoolEntry.platformId,
				)

				time.Sleep(time.Second * time.Duration(scraperPoolEntry.config.Frequency))
			}
		}()
	}
	select {}
}

type RepostReaderService interface {
	PickUpMostTrending(clearUp bool) []domain.Repost
	PurgeIrrelevant() error
}
type RepostApiMessage struct {
	PublicationId string `json:"publication-id"`
	PostedAt      string `json:"posted-at"`
	RepostedAt    string `json:"reposted-at"`
}

func runApi(store repost.Store) {
	var service RepostReaderService
	service = repost.NewService(store)
	r := mux.NewRouter()
	r.HandleFunc("/reposts", func(response http.ResponseWriter, request *http.Request) {
		response.WriteHeader(200)

		reposts := service.PickUpMostTrending(true)
		repostMessages := make([]RepostApiMessage, len(reposts))
		for k, r := range reposts {
			repostMessages[k] = RepostApiMessage{
				string(r.Pub.Id),
				r.Pub.PostedAt.String(),
				r.RepostedAt.String(),
			}
		}
		jsonOutput, _ := json.Marshal(repostMessages)
		_, err := response.Write(jsonOutput)

		if err != nil {
			log.Error("Error sending API response")
		}
		go func() {
			if err := service.PurgeIrrelevant(); err != nil {
				log.Errorf("cannot purge irrelevant reposts: %s", err.Error())
			}
		}()
	})
	log.Info("Running HTTP API on port 35971")

	go func() {
		err := http.ListenAndServe(":"+strconv.Itoa(35971), r)
		if err != nil {
			log.Fatalf("failed to run API server")
		}
	}()
}
